src/Maestro/Maestro.Web/Api/v2019_01_16/Controllers/GoalController.cs

```diff
@@ -0,0 +1,97 @@
+        ///   Exposes methods to set goal time <see cref="Goal"/>
+        /// </summary>
+        /// <param name="ChannelName"></param>
+        /// <param name="DefinitionId"></param>
+        /// <param name="Minutes"></param>
+        [HttpPost]
+        [SwaggerApiResponse(System.Net.HttpStatusCode.Created, Type = typeof(Models.Goal), Description = "Goal for given Channel and Definition-Id is created/updated")]
+        [ValidateModelState]
+        public virtual async Task<IActionResult> Create([Required] int DefinitionId, [Required] string ChannelName, [Required] int Minutes)
+        {
+            Data.Models.Channel channel = await _context.Channels
+                    .Where(c => c.Name.Equals(ChannelName)).FirstOrDefaultAsync();
+            Data.Models.GoalTime goal = await _context.GoalTime
+                .Where(g => g.DefinitionId == DefinitionId && g.ChannelId == channel.Id).FirstOrDefaultAsync();
+
+            if (channel == null)
+            {
+                return NotFound();
+            }
+            if (goal == null)
+            {
+                var goalModel = new Data.Models.GoalTime
+                {
+                    DefinitionId = DefinitionId,
+                    Minutes = Minutes,
+                    ChannelId = channel.Id
+                };
+                await _context.GoalTime.AddAsync(goalModel);
+                await _context.SaveChangesAsync();
+                var goalTest = CreatedAtRoute(
+                    new
+                    {
+                        action = "CreateGoal",
+                        DefinitionId = DefinitionId,
+                        Minutes = Minutes,
+                        ChannelId = channel.Id
+                    },
+                    new Goal(goalModel));
+                return StatusCode((int)HttpStatusCode.Created);
+            }
+            else
+            {
+                goal.Minutes = Minutes;
+                _context.GoalTime.Update(goal);
+                await _context.SaveChangesAsync();
+
+                return Ok(new Goal(goal));
+            }
+        }
+
+        /// <summary>
+        /// Get goal time in minutes <see cref="Goal"/>s that match the given classification.
+        /// </summary>
+        /// <param name="ChannelName"></param>
+        /// <param name="DefinitionId"></param>
+        [HttpGet]
+        [SwaggerApiResponse(HttpStatusCode.OK, Type = typeof(Goal), Description = "Get the Goal for given Channel and Definition-Id")]
+        [ValidateModelState]
+        public virtual async Task<IActionResult> GetGoalTimes([Required]int DefinitionId, [Required]string ChannelName)
+        {
+            Data.Models.Channel channel = await _context.Channels
+                    .Where(c => c.Name.Equals(ChannelName)).FirstOrDefaultAsync();
+            Data.Models.GoalTime goal = await _context.GoalTime
+            .Where(g => g.DefinitionId == DefinitionId && g.ChannelId == channel.Id).FirstOrDefaultAsync();
```

Comment:
You call "FirstOrDefault" above, but then unconditionally reference the value here, which will result in a null reference exception.

Is it possible to do this with one LINQ-TO-SQL request, so that we don't have to make 2 requests to the server (by invoking FirstOrDefaultAsync)? @alexperovich might have an idea how to do that. I'd know how to do it in SQL, not sure how to in EF.

Reply:
It can be done with a Join, but the syntax for that is super weird. There isn't a huge perf hit with 2 queries, its all done through the same connection. The Where should be combined into the FirstOrDefaultAsync calls though. All the First/Single related functions take a predicate just like Where.

Reply:
This isn't resolved either. If "channel" is null (because I sent "PizzaPizzaPizza" as my request), this will crash with a NullReferenceException. And, per @alexperovich comment `.Where(predicate).FirstOrDefaultAsync()` can be more concisely written as `.FirstOrDefaultAsync(predicate)`

Reply:
@ChadNedzlek I thought this was referencing on the EF thingy. I got the null reference thingy now. I will send a not found incase its null.